File: scala/linked-list/src/main/scala/Deque.scala
class Deque[T] {
  class Node(val value: T, var next: Option[Node], var prev: Option[Node]) {}

  var first: Option[Node] = None
  var last: Option[Node] = None

  def addToEmpty(newValue: T) = {
    val newElement = new Node(newValue, None, None)
    first = Some(newElement)
    last = Some(newElement)
  }

  def push(value: T) = last match {
    case Some(x) =>
      last = Some(new Node(value, x.next, last))
      x.next = last
    case None => addToEmpty(value)
  }

  def pop(): Option[T] = last match {
    case Some(x) =>
      last = x.prev
      Some(x.value)
    case None => None
  }

  def shift(): Option[T] = first match {
    case Some(x) =>
      first = x.next
      Some(x.value)
    case None => None
  }

  def unshift(value: T) = first match {
    case Some(x) =>
      first = Some(new Node(value, first, x.prev))
      x.prev = first
    case None => addToEmpty(value)
  }
}

object Deque {
  def apply[T]() = new Deque[T]
}

File: java/linked-list/src/main/java/Deque.java
public class Deque<T> {

    private class Node {
        private T value;
        private Node next;
        private Node prev;

        Node(T value, Node next, Node prev) {
            this.value = value;
            this.next = next;
            this.prev = prev;
        }
    }

    private Node first;
    private Node last;

    public void push(T value) {
        Node newLast = new Node(value, null, this.last);

        if (this.last != null) {
            this.last.next = newLast;
        }

        this.last = newLast;

        if (this.first == null) {
            this.first = this.last;
        }
    }

    public T pop() {
        T value = this.last.value;
        this.last = this.last.prev;

        if (this.last != null) {
            this.last.next = null;
        }

        return value;
    }

    public void unshift(T value) {
        Node newFirst = new Node(value, this.first, null);

        if (this.first != null) {
            this.first.prev = newFirst;
        }

        this.first = newFirst;

        if (this.last == null) {
            this.last = this.first;
        }
    }

    public T shift() {
        T value = this.first.value;
        this.first.prev = null;
        this.first = this.first.next;

        return value;
    }
}